### ISSUE TITLE
Minimal API Support

### ISSUE THREAD

**Issue by mumby0168** [jbogard/MediatR#653]
Seeing the recent advancements on the minimal APIs within ASP.NET Core. 

It would be really cool if we could look at some extension methods that sit on top of this and provide a way to map `IRequest<T>` handlers directly to http routes without having to use a controller. 

I expect this would be a set of extension methods.

**Comment by elementh** [comment]
That would be very cool, but my opinion is that it should be in the form of an extension nuget package, maybe?

**Comment by mumby0168** [comment]
Yeah something like `MediatR.Extensions.AspNetCore.Http` ?

I would be happy to start cracking with this would this be another repository you would look to create?

**Comment by mumby0168** [comment]
I started to put together a project to scope this out you can see it here https://github.com/mumby0168/blog-samples/tree/main/new-features/MinimalApis

While doing so I came across a current limitation of the minimal API's I have logged in issue about it here: https://github.com/dotnet/aspnetcore/issues/35304

**Comment by mumby0168** [comment]
@jbogard what do you think about this? They have recently added a more flexible way to allow custom parameters to be bound. See the previous comment with a link to the issue.

**Comment by Kahbazi** [comment]
> Yeah something like MediatR.Extensions.AspNetCore.Http ?

@mumby0168 This is what I have developed in https://github.com/Kahbazi/MediatR.AspNetCore.Endpoints

You can register `IRequest<T>` directly to the ASP.NET Core routing without Controllers.

```csharp
public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddMediatR(GetType().Assembly);
        services.AddMediatREndpoints();
    }

    public void Configure(IApplicationBuilder app)
    {
        app.UseRouting();

        app.UseEndpoints(endpoints =>
        {
            endpoints.MapMediatR();
        });
    }
}
```

**Comment by mumby0168** [comment]
> > Yeah something like MediatR.Extensions.AspNetCore.Http ?
> 
> @mumby0168 This is what I have developed in https://github.com/Kahbazi/MediatR.AspNetCore.Endpoints
> 
> You can register `IRequest<T>` directly to the ASP.NET Core routing without Controllers.
> 
> ```cs
> public class Startup
> {
>     public void ConfigureServices(IServiceCollection services)
>     {
>         services.AddMediatR(GetType().Assembly);
>         services.AddMediatREndpoints();
>     }
> 
>     public void Configure(IApplicationBuilder app)
>     {
>         app.UseRouting();
> 
>         app.UseEndpoints(endpoints =>
>         {
>             endpoints.MapMediatR();
>         });
>     }
> }
> ```

Ah that is really cool, how do you handle mapping query params to an object?

**Comment by bobbyangers** [comment]
Yeah that us pretty cool.  And how would you apply security requirements?

**Comment by Kahbazi** [comment]
> Ah that is really cool, how do you handle mapping query params to an object?

I haven't implement that yet. 

> Yeah that us pretty cool. And how would you apply security requirements, on the IRequest with a custom filter ?

You can apply `[Authorize]` attributes the same way as controllers and it will work as long as you also add authorization middleware.

**Comment by srollinet** [comment]
I try to do something similar, with extension methods that allow to map a route to a an IRequest.

Something like that

```csharp
app.MediatR().MapGet("/hello", (string name) => new HelloQuery(name));
```

Which will be the equivalent of

```csharp
app.MapGet("/hello", (string name, IMediator mediator, CancellationToken token) =>
     mediator.Send(new HelloQuery(name), token)
);
```

The idea is to have something similar as the original API, without less boilerplate.

The problem is that I don't know the correct approach to use... I tried to create a dynamic delegate that wrap the original one and that adds the mediator and the token, but I don't think it is a good approach, and I have a lack of knowledge in this domain so I struggled with lot of problems.

Maybe the most simple solution is to have a middleware that handles the IRequest, but with this solution, I don't know how Swagger will know the actual return type of the method. I will try to dig a bit.
